### ISSUE TITLE
TestMatches  Test failing with Golang 1.16-beta1

### ISSUE THREAD

**Issue by olivierlemasle** [moby/moby#41953]
**Description**

Test `TestMatches` in `./pkg/fileutils/` is failing with Go 1.16-beta1.
It works correctly with Go <= 1.15

**Steps to reproduce the issue:**

```
go test -v ./pkg/fileutils/
```

**Describe the results you received:**

```
=== RUN   TestCopyFileWithInvalidSrc
--- PASS: TestCopyFileWithInvalidSrc (0.00s)
=== RUN   TestCopyFileWithInvalidDest
--- PASS: TestCopyFileWithInvalidDest (0.00s)
=== RUN   TestCopyFileWithSameSrcAndDest
--- PASS: TestCopyFileWithSameSrcAndDest (0.00s)
=== RUN   TestCopyFileWithSameSrcAndDestWithPathNameDifferent
--- PASS: TestCopyFileWithSameSrcAndDestWithPathNameDifferent (0.00s)
=== RUN   TestCopyFile
--- PASS: TestCopyFile (0.00s)
=== RUN   TestReadSymlinkedDirectoryExistingDirectory
--- PASS: TestReadSymlinkedDirectoryExistingDirectory (0.00s)
=== RUN   TestReadSymlinkedDirectoryNonExistingSymlink
--- PASS: TestReadSymlinkedDirectoryNonExistingSymlink (0.00s)
=== RUN   TestReadSymlinkedDirectoryToFile
--- PASS: TestReadSymlinkedDirectoryToFile (0.00s)
=== RUN   TestWildcardMatches
--- PASS: TestWildcardMatches (0.00s)
=== RUN   TestPatternMatches
--- PASS: TestPatternMatches (0.00s)
=== RUN   TestExclusionPatternMatchesPatternBefore
--- PASS: TestExclusionPatternMatchesPatternBefore (0.00s)
=== RUN   TestPatternMatchesFolderExclusions
--- PASS: TestPatternMatchesFolderExclusions (0.00s)
=== RUN   TestPatternMatchesFolderWithSlashExclusions
--- PASS: TestPatternMatchesFolderWithSlashExclusions (0.00s)
=== RUN   TestPatternMatchesFolderWildcardExclusions
--- PASS: TestPatternMatchesFolderWildcardExclusions (0.00s)
=== RUN   TestExclusionPatternMatchesPatternAfter
--- PASS: TestExclusionPatternMatchesPatternAfter (0.00s)
=== RUN   TestExclusionPatternMatchesWholeDirectory
--- PASS: TestExclusionPatternMatchesWholeDirectory (0.00s)
=== RUN   TestSingleExclamationError
--- PASS: TestSingleExclamationError (0.00s)
=== RUN   TestMatchesWithNoPatterns
--- PASS: TestMatchesWithNoPatterns (0.00s)
=== RUN   TestMatchesWithMalformedPatterns
--- PASS: TestMatchesWithMalformedPatterns (0.00s)
=== RUN   TestMatches
    fileutils_test.go:387: assertion failed: error is not nil: syntax error in pattern: pattern="a\\" text="a\\"
--- FAIL: TestMatches (0.00s)
=== RUN   TestCleanPatterns
--- PASS: TestCleanPatterns (0.00s)
=== RUN   TestCleanPatternsStripEmptyPatterns
--- PASS: TestCleanPatternsStripEmptyPatterns (0.00s)
=== RUN   TestCleanPatternsExceptionFlag
--- PASS: TestCleanPatternsExceptionFlag (0.00s)
=== RUN   TestCleanPatternsLeadingSpaceTrimmed
--- PASS: TestCleanPatternsLeadingSpaceTrimmed (0.00s)
=== RUN   TestCleanPatternsTrailingSpaceTrimmed
--- PASS: TestCleanPatternsTrailingSpaceTrimmed (0.00s)
=== RUN   TestCleanPatternsErrorSingleException
--- PASS: TestCleanPatternsErrorSingleException (0.00s)
=== RUN   TestCreateIfNotExistsDir
--- PASS: TestCreateIfNotExistsDir (0.00s)
=== RUN   TestCreateIfNotExistsFile
--- PASS: TestCreateIfNotExistsFile (0.00s)
=== RUN   TestMatch
--- PASS: TestMatch (0.00s)
FAIL
FAIL	github.com/docker/docker/pkg/fileutils	0.011s
FAIL

```

The offending lines: https://github.com/moby/moby/blob/3e0025e2fc137f51e95758618e1b42627fe6ea1c/pkg/fileutils/fileutils_test.go#L380-L381

(note that this test is already disabled on windows, because it was failing).

**Comment by thaJeztah** [comment]
Odd; trying to reproduce with the following, but don't see it failing (test could be simplified even further);

```go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

func main() {
	_, err := NewPatternMatcher([]string{"a\\\\"})
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("success")
	}
}

// Pattern defines a single regexp used to filter file paths.
type Pattern struct {
	cleanedPattern string
	dirs           []string
	regexp         *regexp.Regexp
	exclusion      bool
}

// PatternMatcher allows checking paths against a list of patterns
type PatternMatcher struct {
	patterns   []*Pattern
	exclusions bool
}

// NewPatternMatcher creates a new matcher object for specific patterns that can
// be used later to match against patterns against paths
func NewPatternMatcher(patterns []string) (*PatternMatcher, error) {
	pm := &PatternMatcher{
		patterns: make([]*Pattern, 0, len(patterns)),
	}
	for _, p := range patterns {
		// Eliminate leading and trailing whitespace.
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		p = filepath.Clean(p)
		newp := &Pattern{}
		if p[0] == '!' {
			if len(p) == 1 {
				return nil, errors.New("illegal exclusion pattern: \"!\"")
			}
			newp.exclusion = true
			p = p[1:]
			pm.exclusions = true
		}
		// Do some syntax checking on the pattern.
		// filepath's Match() has some really weird rules that are inconsistent
		// so instead of trying to dup their logic, just call Match() for its
		// error state and if there is an error in the pattern return it.
		// If this becomes an issue we can remove this since its really only
		// needed in the error (syntax) case - which isn't really critical.
		if _, err := filepath.Match(p, "."); err != nil {
			return nil, err
		}
		newp.cleanedPattern = p
		newp.dirs = strings.Split(p, string(os.PathSeparator))
		pm.patterns = append(pm.patterns, newp)
	}
	return pm, nil
}
```

**Comment by thaJeztah** [comment]
oh! testing the wrong case (`a\\\\` instead of `a\\`); let me try again

**Comment by thaJeztah** [comment]
Yes, can confirm that that fails;

```go
package main

import (
        "fmt"
        "path/filepath"
)

func main() {
        p := filepath.Clean("a\\")
        fmt.Println("cleaned:", p)

        _, err := filepath.Match(p, ".")

        if err != nil {
                fmt.Println(err)
        } else {
                fmt.Println("success")
        }
}
```


```bash
docker run -it --rm -v (pwd):/foo golang:1.15-alpine sh -c 'go run /foo/main.go'
cleaned: a\
success
```

```bash
docker run -it --rm -v (pwd):/foo golang:1.16beta1-alpine sh -c 'go run /foo/main.go'
cleaned: a\
syntax error in pattern
````

**Comment by thaJeztah** [comment]
Could be related to https://tip.golang.org/doc/go1.16#path/filepath

> The Match and Glob functions now return an error if the unmatched part of the pattern has a syntax error. Previously, the functions returned early on a failed match, and thus did not report any later syntax error in the pattern.

**Comment by olivierlemasle** [comment]
Yes, it seems to be directly related to the change you mentioned above.

It seems that removing the 2 offending test cases is sufficient, no? No need to fix something else in the `NewPatternMatcher` logic?

**Comment by thaJeztah** [comment]
Fixing the test could be an option; mostly curious what exactly it was testing, and if there will be a change in behavior for end-users 🤔
